### PR TITLE
Add developer document describing system architecture

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Adds a refreshed and more detailed document dscribing the architecture of the codebase. It is a more modern and updated version of the original [ADD](https://github.com/mantidproject/documents/blob/a8e584151871bb61b4db2497af2e4cd5dc0f32da/Design/ArchitectureDesignDocument.doc).

**To test:**

Review the document and please make suggestions for improving it's clarity.

*There is no associated issue.*

*This does not require release notes* because **it is developer documentation**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: dev-docs/source/Architecture.rst [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+Architecture
+============
+
+.. contents::
+   :local:
+
+Summary
+-------
+
+Mantid is used for reduction and analysis of neutron and muon data.
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:11]: I'm in favour of not forgetting the data visualization aspect, I think it's important to put it in the opening sentence.

File: dev-docs/source/Architecture.rst [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+by the following context diagram.
+
+.. figure:: /images/architecture-context.png
+   :scale: 70%
+   :align: center
+   :alt: Context diagram of mantid software system
+
+   Context diagram for mantid software
+
+Experimental users of at neutron/muon facilities use the software
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:28]: trailing `of`

File: dev-docs/source/Architecture.rst [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+Container
+---------
+
+*Please note that container here does not mean container in the
+deployment (Docker, Singularity) sense of the word, rather the generic
+definition of enclosing something.*
+
+The mantid software system is defined as four containers:
+two libraries (``framework`` & ``mantidqt``),
+one general-purpose graphical application (``workbench``) and
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:46]:
```suggestion
one general-purpose, cross-platform, graphical desktop application
```

File: dev-docs/source/Architecture.rst [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+widgets that are composed to form the ``workbench`` application.
+This library is accessible outside of the workbench application to allow
+custom applications to be built from the widgets.
+
+Please note the directional aspect of the connection/dependency arrows.
+The connections are strictly one way and those earlier in the dependency chain
+cannot contain dependencies on those later, e.g.
+
+- framework *does not* depend on ``mantidqt`` or ``workbench``
+- ``mantidqt`` *does not* depend on ``workbench``
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:72]:
```suggestion
- ``mantidqt`` *does not* depend on ``workbench`` but does depend on ``framework``
```

File: dev-docs/source/Architecture.rst [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+########
+
+.. figure:: /images/architecture-mantidqt.png
+   :scale: 70%
+   :align: center
+   :alt: Component diagram of mantidqt container
+
+   Container diagram for mantidqt Python package
+
+``mantidqt`` is a Python package defined house a collcetion of Qt widgets
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:123]: there are typos here

File: dev-docs/source/Architecture.rst [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+   :scale: 70%
+   :align: center
+   :alt: Component diagram of workbench container
+
+   Container diagram for workbench application
+
+``workbench`` is an application written in pure Python in the ``qt/applications/workbench``
+directory of the source.
+It is not intended to be a library imported by others and as such offers
+so guarantees of API stability.
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:159]:
```suggestion
no guarantees of API stability.
```

File: dev-docs/source/Architecture.rst [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+It makes heavy use of the ``mantidqt`` widgets collection and puts them
+together to build the application shipped to users as ``MantidWorkbench``.
+
+The application provides some widgets that are only applicable to itself,
+such as the about screen, settings screen and project recovery mechanism that
+is specific to workbench.
+
+Plotting within ``workbench`` uses matplotlib_.
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:167]: Shall we mention somewhere the OpenGL used in the instrument viewer? Or perhaps a small paragraph can mention that apart from plotting with matplotlib there are 2 advanced widgets: sliceviewer and instrument viewer.

File: dev-docs/source/Architecture.rst [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+This is illustrated, along with connections between the containers,
+in the following diagram.
+
+
+.. figure:: /images/architecture-container.png
+   :scale: 70%
+   :align: center
+   :alt: Container diagram of mantid software system
+
+   Container diagram for mantid software
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:57]: It's unclear if the workbench depends on the custom interfaces. They are not really necessary for workbench, still they are built when building the workbench and appear in the menus.

**Review by gvardany — COMMENTED** [review]
Thanks for turning the ADD into a living document, this reflects how the project is organized currently.
Could you perhaps make the diagrams a little larger, maybe only the first two? The text was barely readable on those.

File: dev-docs/source/Architecture.rst (hunk 1 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:11]: Good point. We spend a lot of time writing those!

File: dev-docs/source/Architecture.rst (hunk 5 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:123]: Thank you. Looking back it's clear I didn't read that part over again!

File: dev-docs/source/Architecture.rst (hunk 7 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:167]: I might do this in the `mantidqt` and emphasize that it's not just basic thinks like text fields and combo boxes.

File: dev-docs/source/Architecture.rst (hunk 8 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:57]: They are not strict dependencies from the code side, they are basically plugins as we do the registration and discover at run time. I will try to make this clearer in the diagram.

**Review by martyngigg — COMMENTED** [review]
Thanks for the comments @gvardany and sorry for the delay in getting back to you with the comments.

File: dev-docs/source/Architecture.rst [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+Container
+---------
+
+*Please note that container here does not mean container in the
+deployment (Docker, Singularity) sense of the word, rather the generic
+definition of enclosing something.*
+
+The mantid software system is defined as four containers:
+two libraries (``framework`` & ``mantidqt``),
+one general-purpose, cross-platform, graphical desktop application
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:45]: Is there a comma missing here?

File: dev-docs/source/Architecture.rst [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+Plotting within ``workbench`` uses matplotlib_.
+The application bundle provides some custom features to step in on the backend,
+in ``workbench.plotting``, so that figures displayed within workbench offer
+a richer experience than a standard matplotlib figure,
+e.g. interactive fitting to data within a figure.
+The aim of these customizations is to be transparent to a user in terms of the
+matplotlib api.
+A plotting script should function, without modification,
+outside of workbench with the standard matplotlib backends,
```
Comment (gvardany) [inline dev-docs/source/Architecture.rst:181]:
```suggestion
outside of workbench with the standard matplotlib backends, including jupyter notebooks
```
Perhaps we could also make explicit in the Framework/Python API section, that mantid as a framework can be used in vanilla python in batch mode or in any interactive python environment, including jupyter notebooks.

File: dev-docs/source/images/architecture-container.drawio [hunk 11]
```diff
@@ -0,0 +1 @@
...ZiSB6XVwp/EZ6q2QCElkAkrcONYZhFnpRK+DlRQu5bmRcImviAi/OVlMxm4EorW0Al0G+NwmyAtTnMxZ6CTnMlsai30HlS4wlVWtxkLD+iLm8xlTA7kZIrRqhpiT+iYT1sllBlMfLhw9kEOQuLYJkJAWhPV2Ao4UQ1XTqDQYuXCG83MkavNkQAGeOsE0WbhBHDLz6yMqMFis9BmnAEIS+T6Ehii0ob6awEDvAxD78gVG4Udg3Ew/6KSFfL8UHxqBaTj8u5OtHQfwPWMrRWoYZynHCDNUsOQvliPHQYYZ0FEH5BK5BW8E5sYx/NT1VE0ZmocTHCyjTrV1d+A+KphU3lOyZkTy8/1Ddw3oZBgS5lJDMZt/lI7CqW1CIyWbUPCrSI7QWnv0SRob/YOHnG5OP/xfBo2a8Y6ZRKFxt9rZ70wt2nIWBACOsbRWpXUcvwUYz9tnXXqT5y51rU1vsQ+Mexc4MRH4rpoDhSdueDsigX3Jd9ZIgxQuvOCCVgfRS2sNUjuzJ9CuC+j3dTsl3vz0FUeDysVKojVrZicHTTe3UTZzR6Ka2qXYXkP09N628DQlbAVLj0ZatgPRuySzhbMwGy94FH/fkx65W5gh8PFZlrLT70FT7xwX66hmXZ0y9PYbz46jU/eGqZE2K0AnEa32lwLNEVk4X8LckJo6C+IkTmnNBM4dU7ISePtpr74X24vxE9VBWnJ+Qbp/uYlGnxHIEGl2XdRammU9Xy7tLXUjnat/cU6/TGvohah9CATAq/O/mFOz2EyH5U6wVgaG8IGcNp6J4YlTwAwHtZUqUoygiWC0ctE8olPLLpqAYne5uKpzDPNTOpWqt8gvwiM0SVRnDTpIbrXPJbzjwUzzvY6pmO+508RsPYueq/nga10crNVuZyiUnAXNNLfRWoXobKWxJdXmwfoGtRXR2rZ3GBgvgs6/WxrEp07C6PQs4idk7mqQNwFrOI6bfh4QOAxNp41zXBTh/wONgWl8riEe/X1Xe5IhJoarYgXSInYw2R6i3JAPfrWYYXT1EFekICvYsJAq2ijRq1FaHK8sAn/e5Rxh9ijKgadq52c8dUzWqItGQLWwf5mlqemY/ET9tx9QgG6phSYMymh06gKz6HFd2sKQtd/en6Tk2K+PunuOemYBesU5WSgPs4jPSy+xHp+Lu2U93qdf/Ag==</diagram></mxfile>
```
Comment (gvardany) [inline dev-docs/source/images/architecture-container.drawio:1]: Github marks the io files not having a new line at the end, I don't think this is critical though.

**Review by gvardany — COMMENTED** [review]
Thanks for the changes, just a couple of more remarks:

File: dev-docs/source/Architecture.rst (hunk 9 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:45]: I'm missing an `and` I think.

File: dev-docs/source/images/architecture-container.drawio (hunk 11 above)
Comment (martyngigg) [reply dev-docs/source/images/architecture-container.drawio:1]: I'd prefer to keep those as saved by drawio as if it's updated then it'll just get overwritten.

File: dev-docs/source/Architecture.rst (hunk 10 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:181]: Yes that's true. Being more explicit is worth it here.

**Comment by martyngigg** [comment]
Thanks, Gagik. I've addressed those few issues.

**Review by gvardany — DISMISSED** [review]

File: dev-docs/source/Architecture.rst [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+-------
+
+Mantid is used for reduction, analysis & visualization of neutron and muon data.
+This document describes the architecture of the mantid software in terms
+of the `C4 model <C4Model_>`_.
+
+Context
+-------
+
+The mantid software interacts with people and other systems as described
```
Comment (DavidFair) [inline dev-docs/source/Architecture.rst:18]: Nit this sentence is a bit clunky
```suggestion
People and other systems interact with the software package called Mantid as described
```

File: dev-docs/source/Architecture.rst [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+.. figure:: /images/architecture-workbench.png
+   :scale: 85%
+   :align: center
+   :alt: Component diagram of workbench container
+
+   Container diagram for workbench application
+
+``workbench`` is an application written in pure Python in the ``qt/applications/workbench``
+directory of the source.
```
Comment (DavidFair) [inline dev-docs/source/Architecture.rst:169]: Reading through we don't explicitly state that Workbench is our main "entrypoint" which allows users to do the majority of their work from a GUI perspective (that isn't an external GUI). I'm not sure if we want to state it here alongside the library section, or above where we introduce this?

```suggestion
directory of the source. It is intended to be the main interface provided by the project and allows access to the customized interfaces as described in `Container<container_>`.
```
Feel free to adapt this further as I'm not 100% how to phrase it

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor nits on clarity. Otherwise builds correctly for me, renders correctly and reads well

File: dev-docs/source/Architecture.rst (hunk 13 above)
Comment (martyngigg) [reply dev-docs/source/Architecture.rst:169]: Yeah this is a good point. Developers reading this should have an understanding of how the user community uses things.

**Comment by martyngigg** [comment]
Thanks, David. I've addressed those couple of points.

**Review by DavidFair — APPROVED** [review]
Thanks for doing those nit picks :+1:

**Pull request merged**